util/topsql/reporter/pubsub.go

```diff
@@ -0,0 +1,257 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reporter
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tipb/go-tipb"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+// TopSQLPubSubService implements tipb.TopSQLPubSubServer.
+//
+// If a client subscribes to TopSQL records, the TopSQLPubSubService is responsible
+// for registering an associated DataSink to the reporter. Then the reporter sends
+// data to the client periodically.
+type TopSQLPubSubService struct {
+	dataSinkRegHandle DataSinkRegHandle
+}
+
+// NewTopSQLPubSubService creates a new TopSQLPubSubService.
+func NewTopSQLPubSubService(
+	dataSinkRegHandle DataSinkRegHandle,
+) *TopSQLPubSubService {
+	return &TopSQLPubSubService{
+		dataSinkRegHandle: dataSinkRegHandle,
+	}
+}
+
+var _ tipb.TopSQLPubSubServer = &TopSQLPubSubService{}
+
+// Subscribe registers dataSinks to the reporter and redirects data received from reporter
+// to subscribers associated with those dataSinks.
+func (t *TopSQLPubSubService) Subscribe(
+	_ *tipb.TopSQLSubRequest,
+	stream tipb.TopSQLPubSub_SubscribeServer,
+) error {
+	ds := newPubSubDataSink(stream)
+	t.dataSinkRegHandle.Register(ds)
+	ds.run()
+	return nil
+}
+
+type pubSubDataSink struct {
+	stream     tipb.TopSQLPubSub_SubscribeServer
+	sendTaskCh chan sendTask
+	isDown     *atomic.Bool
+}
+
+func newPubSubDataSink(
+	stream tipb.TopSQLPubSub_SubscribeServer,
+) *pubSubDataSink {
+	return &pubSubDataSink{
+		stream:     stream,
+		sendTaskCh: make(chan sendTask, 1),
+		isDown:     atomic.NewBool(false),
+	}
+}
+
+var _ DataSink = &pubSubDataSink{}
+
+func (s *pubSubDataSink) Send(data ReportData, deadline time.Time) {
+	if s.IsDown() {
+		return
+	}
+
+	select {
+	case s.sendTaskCh <- sendTask{data: data, deadline: deadline}:
+		// sent successfully
+	default:
+		ignoreReportChannelFullCounter.Inc()
+		logutil.BgLogger().Warn("[top-sql] report channel is full")
+	}
+}
+
+func (s *pubSubDataSink) IsPaused() bool {
+	return false
+}
+
+func (s *pubSubDataSink) IsDown() bool {
+	return s.isDown.Load()
+}
+
+func (s *pubSubDataSink) Close() {
+	close(s.sendTaskCh)
+}
+
+func (s *pubSubDataSink) run() {
+	defer s.isDown.Store(true)
+
+	for task := range s.sendTaskCh {
+		ctx, cancel := context.WithDeadline(context.Background(), task.deadline)
+		start := time.Now()
+
+		var err error
+		go util.WithRecovery(func() {
+			defer cancel()
+			err = s.doSend(ctx, task.data)
+			if err != nil {
+				reportAllDurationFailedHistogram.Observe(time.Since(start).Seconds())
+			} else {
+				reportAllDurationSuccHistogram.Observe(time.Since(start).Seconds())
+			}
+		}, nil)
+
+		<-ctx.Done()
```

Comment:
If you are waiting goroutine to finish, why not execute it in current goroutine directly?

Reply:
That spawned goroutine may block when the deadline is exceeded.

Reply:
Got it. You can add a comment to document the reason.